debug/debugger.go

```diff
@@ -0,0 +1,859 @@
+	}
+
+	t, err := s.thread(threadID)
+	if err != nil {
+		return nil, err
+	}
+
+	threadFrames := s.framesByThread[t.id]
+	if threadFrames == nil {
+		threadFrames = []*stackFrame{}
+	}
+
+	stackIndex := 0
+	if len(threadFrames) > 0 {
+		stackIndex = threadFrames[len(threadFrames)-1].stackIndex + 1
+	}
+	newEvents := t.stackEvents(stackIndex)
+	for _, e := range newEvents {
+		info := s.newStackFrame(e, t, stackIndex)
+		stackIndex++
+		threadFrames = append(threadFrames, info)
+	}
+	s.framesByThread[t.id] = threadFrames
+
+	frames := make([]StackFrame, 0, len(threadFrames))
+	for _, f := range threadFrames {
+		frames = append(frames, f)
+	}
+	slices.Reverse(frames)
+
+	return frames, nil
+}
+
+func (s *session) newStackFrame(e *topdown.Event, t *thread, stackIndex int) *stackFrame {
+	id := len(s.frames) + 1 // frames are 1-indexed
+
+	var expl string
+	if e.Node != nil {
+		pretty := new(bytes.Buffer)
+		topdown.PrettyTrace(pretty, []*topdown.Event{e})
+		expl = strings.Trim(pretty.String(), "\n")
+	} else {
+		expl = fmt.Sprintf("%s, %s", e.Op, e.Location)
+	}
+
+	frame := &stackFrame{
+		id:         FrameID(id),
+		name:       fmt.Sprintf("#%d: %d %s", id, e.QueryID, expl),
+		location:   e.Location,
+		thread:     t.id,
+		stackIndex: stackIndex,
+		e:          e,
+	}
+
+	s.frames = append(s.frames, frame)
+	return frame
+}
+
+func (s *session) frame(id FrameID) (*stackFrame, error) {
+	index := int(id) - 1
+	if index < 0 || index >= len(s.frames) {
+		return nil, fmt.Errorf("invalid frame id: %d", id)
+	}
+	return s.frames[index], nil
+}
+
+func (s *session) Scopes(frameID FrameID) ([]Scope, error) {
+	if s == nil {
+		return nil, fmt.Errorf("no active debug session")
+	}
+
+	f, err := s.frame(frameID)
+	if err != nil {
+		return nil, err
+	}
+
+	t, err := s.thread(f.thread)
+	if err != nil {
+		return nil, err
+	}
+
+	return t.scopes(f.stackIndex), nil
+}
+
+func (s *session) Variables(varRef VarRef) ([]Variable, error) {
+	if s == nil {
+		return nil, fmt.Errorf("no active debug session")
+	}
+
+	s.d.logger.Debug("Variables requested: %d", varRef)
+
+	vars, err := s.varManager.vars(varRef)
+	if err != nil {
+		return nil, err
+	}
+
+	return vars, nil
+}
+
+func (s *session) Breakpoints() ([]Breakpoint, error) {
+	if s == nil {
+		return nil, fmt.Errorf("no active debug session")
+	}
+
+	return s.breakpoints.all(), nil
+}
+
+func (s *session) SetBreakpoints(locations []location.Location) ([]Breakpoint, error) {
+	if s == nil {
+		return nil, fmt.Errorf("no active debug session")
+	}
+
+	s.d.logger.Debug("Clearing existing breakpoints")
+	s.breakpoints.clear()
+
+	bps := make([]Breakpoint, 0, len(locations))
+	for _, loc := range locations {
+		bps = append(bps, s.breakpoints.add(loc))
+	}
+
+	s.d.logger.Info("Breakpoints set: %s", bps)
+
+	return bps, nil
+}
+
+func (s *session) AddBreakpoint(loc location.Location) (Breakpoint, error) {
+	if s == nil {
+		return nil, fmt.Errorf("no active debug session")
+	}
```

Comment:
Hmm interesting that `SetBreakpoints` clears existing breakpoints before adding new ones.

Reply:
"Set" implies replacing what already exists, whereas `add` does not.
This was however a misread of the DAP protocol from me, where I originally thought setting breakpoints was a global operation, but it's actually a per-file operation. I solved this by instead making the user of the SDK make that distinction by using a combination of `Breakpoints()`, `AddBreakpoint()`, and `RemoveBreakpoint()`.